In [16]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import cross_val_score, StratifiedKFold
from sklearn.metrics import f1_score

In [4]:
df = pd.read_csv("WA_Fn-UseC_-Telco-Customer-Churn.csv")

df.head()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


In [5]:
df["Partner"] = df["Partner"].fillna(df["Partner"].mode()[0])
df["MonthlyCharges"] = df["MonthlyCharges"].fillna(df["MonthlyCharges"].median())
df['TotalCharges'] = pd.to_numeric(df['TotalCharges'], errors='coerce')
df["TotalCharges"] = df["TotalCharges"].fillna(df["TotalCharges"].median())

In [6]:
df["gender"] = df["gender"].replace({ "Male" : 1, "Female" : 0})
df["Partner"] = df["Partner"].replace({ "Yes" : 1, "No" : 0})
df["Dependents"] = df["Dependents"].replace({ "Yes" : 1, "No" : 0})
df["PhoneService"] = df["PhoneService"].replace({ "Yes" : 1, "No" : 0})
df["MultipleLines"] = df["MultipleLines"].replace({ "Yes" : 1, "No" : 0, "No phone service" : 0 })
df["OnlineSecurity"] = df["OnlineSecurity"].replace({ "Yes" : 1, "No" : 0, "No internet service" : 0 })
df["OnlineBackup"] = df["OnlineBackup"].replace({ "Yes" : 1, "No" : 0, "No internet service" : 0 })
df["DeviceProtection"] = df["DeviceProtection"].replace({ "Yes" : 1, "No" : 0, "No internet service" : 0 })
df["TechSupport"] = df["TechSupport"].replace({ "Yes" : 1, "No" : 0, "No internet service" : 0 })
df["StreamingTV"] = df["StreamingTV"].replace({ "Yes" : 1, "No" : 0, "No internet service" : 0 })
df["StreamingMovies"] = df["StreamingMovies"].replace({ "Yes" : 1, "No" : 0, "No internet service" : 0 })
df["PaperlessBilling"] = df["PaperlessBilling"].replace({ "Yes" : 1, "No" : 0 })
df["StreamingTV"] = df["StreamingTV"].replace({ "Yes" : 1, "No" : 0, "No internet service" : 0 })
df["Churn"] = df["Churn"].replace({"Yes": 1, "No": 0})
df = pd.get_dummies(df, columns=["InternetService", "Contract", "PaymentMethod"])

C:\Users\darag\AppData\Local\Temp\ipykernel_7984\3653414510.py:1: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df["gender"] = df["gender"].replace({ "Male" : 1, "Female" : 0})
C:\Users\darag\AppData\Local\Temp\ipykernel_7984\3653414510.py:2: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df["Partner"] = df["Partner"].replace({ "Yes" : 1, "No" : 0})
C:\Users\darag\AppData\Local\Temp\ipykernel_7984\3653414510.py:3: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the o

In [7]:
X = df.drop(columns=["customerID", "Churn"])
Y = df["Churn"]

In [8]:
X_train, X_test, Y_train, Y_test = train_test_split(
    X, Y, test_size=0.2, random_state=42)

In [9]:
scaler = StandardScaler()

X_train_scaled = pd.DataFrame(
    scaler.fit_transform(X_train),
    columns=X_train.columns,
    index=X_train.index
)

X_test_scaled = pd.DataFrame(
    scaler.transform(X_test),
    columns=X_test.columns,
    index=X_test.index
)

X_train_scaled.head()

,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,OnlineSecurity,OnlineBackup,DeviceProtection,...,InternetService_DSL,InternetService_Fiber optic,InternetService_No,Contract_Month-to-month,Contract_One year,Contract_Two year,PaymentMethod_Bank transfer (automatic),PaymentMethod_Credit card (automatic),PaymentMethod_Electronic check,PaymentMethod_Mailed check
2142,-1.025166,-0.437749,-0.968546,1.532186,-0.465683,0.329573,-0.860523,1.580942,-0.733828,1.384796,...,1.380442,-0.887057,-0.524081,-1.099339,1.910109,-0.558353,-0.530678,-0.527656,-0.709650,1.84247
1623,-1.025166,-0.437749,-0.968546,-0.652662,0.885537,0.329573,1.162083,-0.632534,1.362717,-0.722128,...,-0.724405,1.127324,-0.524081,-1.099339,-0.523530,1.790982,1.884380,-0.527656,-0.709650,-0.54275
6074,0.975452,-0.437749,1.032476,-0.652662,-1.284605,-3.034225,-0.860523,-0.632534,-0.733828,-0.722128,...,1.380442,-0.887057,-0.524081,0.909638,-0.523530,-0.558353,-0.530678,-0.527656,1.409146,-0.54275
1362,0.975452,-0.437749,-0.968546,-0.652662,-1.161766,0.329573,-0.860523,-0.632534,-0.733828,-0.722128,...,-0.724405,1.127324,-0.524081,0.909638,-0.523530,-0.558353,-0.530678,-0.527656,1.409146,-0.54275
6754,0.975452,-0.437749,-0.968546,1.532186,-1.325551,0.329573,1.162083,1.580942,1.362717,-0.722128,...,1.380442,-0.887057,-0.524081,-1.099339,-0.523530,1.790982,1.884380,-0.527656,-0.709650,-0.54275


Step 1: Take a Week 3 model and evaluate it with 5-fold cross-validation using cross_val_score.

In [10]:
model1 = LogisticRegression(C=1, max_iter=1000, class_weight='balanced', random_state=42)
scores = cross_val_score(model1, X_train_scaled, Y_train, cv=5, scoring="f1")
print(scores)

[0.64810127 0.62123198 0.61340206 0.61068702 0.61559508]


Step 2: Report the mean and standard deviation of the scores across folds.

In [11]:
print("Mean =",scores.mean())
print("Std =",scores.std()) 

Mean = 0.6218034809697521
Std = 0.0135985579197657


we can see that 5 scores f1 = [0.648, 0.621, 0.613, 0.610, 0.615] roughly the same and we have a small std, in this case we can say that the model is stable and that's a good sign, so we can trust the 0.622 scores.

Step 3: Compare the cross-validated estimate to the single-split score from Day 1 and explain any 
difference.

In [12]:
X_train2, X_val, Y_train2, Y_val = train_test_split(
    X_train, Y_train, test_size=0.2, random_state=42)

In [13]:
scaler = StandardScaler()

X_train2_scaled = pd.DataFrame(
    scaler.fit_transform(X_train2),
    columns=X_train2.columns,
    index=X_train2.index
)

X_val_scaled = pd.DataFrame(
    scaler.transform(X_val),
    columns=X_val.columns,
    index=X_val.index
)

X_train_scaled.head()

,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,OnlineSecurity,OnlineBackup,DeviceProtection,...,InternetService_DSL,InternetService_Fiber optic,InternetService_No,Contract_Month-to-month,Contract_One year,Contract_Two year,PaymentMethod_Bank transfer (automatic),PaymentMethod_Credit card (automatic),PaymentMethod_Electronic check,PaymentMethod_Mailed check
2142,-1.025166,-0.437749,-0.968546,1.532186,-0.465683,0.329573,-0.860523,1.580942,-0.733828,1.384796,...,1.380442,-0.887057,-0.524081,-1.099339,1.910109,-0.558353,-0.530678,-0.527656,-0.709650,1.84247
1623,-1.025166,-0.437749,-0.968546,-0.652662,0.885537,0.329573,1.162083,-0.632534,1.362717,-0.722128,...,-0.724405,1.127324,-0.524081,-1.099339,-0.523530,1.790982,1.884380,-0.527656,-0.709650,-0.54275
6074,0.975452,-0.437749,1.032476,-0.652662,-1.284605,-3.034225,-0.860523,-0.632534,-0.733828,-0.722128,...,1.380442,-0.887057,-0.524081,0.909638,-0.523530,-0.558353,-0.530678,-0.527656,1.409146,-0.54275
1362,0.975452,-0.437749,-0.968546,-0.652662,-1.161766,0.329573,-0.860523,-0.632534,-0.733828,-0.722128,...,-0.724405,1.127324,-0.524081,0.909638,-0.523530,-0.558353,-0.530678,-0.527656,1.409146,-0.54275
6754,0.975452,-0.437749,-0.968546,1.532186,-1.325551,0.329573,1.162083,1.580942,1.362717,-0.722128,...,1.380442,-0.887057,-0.524081,-1.099339,-0.523530,1.790982,1.884380,-0.527656,-0.709650,-0.54275


In [15]:
best_f1 = 0
best_C = None

for C in [0.1, 1, 10, 100]:
    model2 = LogisticRegression(C=C, max_iter=1000, class_weight='balanced', random_state=42)
    model2.fit(X_train2_scaled, Y_train2)
    pred = model2.predict(X_val_scaled)
    f1 = f1_score(Y_val, pred, pos_label=1)
    print(f"C = {C}, f1 score = {f1:.4f}")
    if f1 > best_f1:
        best_f1 = f1
        best_C = C

print(f"the best C = {best_C}, with f1 score = {best_f1:.4f}")

C = 0.1, f1 score = 0.6292
C = 1, f1 score = 0.6309
C = 10, f1 score = 0.6275
C = 100, f1 score = 0.6275
the best C = 1, with f1 score = 0.6309


the single train/val split gives us a higher f1 (0.6309) than thwe 5-fold cross-val avg (0.6218). in single train/val split the model could get a bit lucky. since the cross-val std is small (0.0135), this difference is just normal variation. I trust the 5-fold cross-val result (0.6218) more, since it's the more reliable estimate.

Step 4: For a classification task, confirm stratified folds are used and explain why that matters here.

In [17]:
skf = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)
scores_skf = cross_val_score(model1, X_train_scaled, Y_train, cv=skf, scoring="f1")
print(scores_skf)

[0.5897772  0.62435233 0.62451613 0.61757106 0.65349144]


In [18]:
print("Mean =",scores_skf.mean())

Mean = 0.6219416302903829


I ran StratifiedKFold and got these scores = [0.5897, 0.6243, 0.6245, 0.6175, 0.6534], scores are very close to the scores from the default cross_val_score(...) used in step1, whoch confirms that cross_val_score already used stratified folds automatically when I passed an integer cv=5 with classifier.
why that matters: this is an imbalanced churn dataset non-stratified folds could end up with very different churn ratios by chance, making the f1 scores unreliable to compare. stratified k-fold keeps the same churn ratioin every fold, so the score is fair, trustworthy estimate.